packages/flutter/lib/src/widgets/app.dart

```diff
@@ -610,6 +613,22 @@ class WidgetsApp extends StatefulWidget {
   /// {@endtemplate}
   final RouteInformationProvider? routeInformationProvider;
 
+  /// {@template flutter.widgets.widgetsApp.routeInformationCodec}
+  /// A codec to convert [RouteInformation] into a serializable form.
+  ///
+  /// This codec is used for encoding and decoding route information during
+  /// router's state restoration.
```

Comment:
I thought we also need to encode the state for the web to push it into the browser's history. Why is the codec not used for that? How is serialization achieved there?

Reply:
normally the report to web is handled directly inside the PlatformRouteinformationprovider or whichever custom subclass customer write.  The router isn't involved in the process. We can update PlatformRouteinformationprovider to also take in a codec to be use when serialize into web.

Reply:
I see, on the web the JSONMethodCodec is used implicitly to encode the state when we call `SystemNavigator.routeInformationUpdated` from `PlatformRouteInformationProvider.routerReportsNewRouteInformation`. People do have the option to override this method and provide their own serialization there, I guess. But as far as I can tell, we provide no hook for custom deserialization for information coming from the engine, so overriding the serialization is kinda pointless?

If we introduce a codec, we should use it consistently for all serialization needs.